apps/scan/backend/src/scanners/custom/state_machine.ts

```diff
@@ -516,6 +522,37 @@ function buildMachine({
     };
   }
 
+  const doScanWhenReady: InvokeConfig<Context, Event> = {
+    src: (context) =>
+      buildPaperStatusObserver(
+        delays.DELAY_PAPER_STATUS_POLLING_INTERVAL,
+        delays.DELAY_PAPER_STATUS_POLLING_TIMEOUT
+      )(context).pipe(
+        switchMap(async (event) => {
+          if (event.type === 'SCANNER_READY_TO_SCAN') {
+            const flowState = await getCurrentAppFlowState({
+              auth,
+              store: workspace.store,
+              usbDrive,
+              // NOTE: we only ever call `doScanWhenReady` when the scanner is
+              // ready to scan, so we can assume that the scanner is ready to
+              // scan.
+              precinctScannerState: 'ready_to_scan',
+            });
+            if (flowState === 'ballot:waiting_to_scan') {
+              return { type: 'SCAN' };
+            }
+          }
+
+          return event;
+        })
```

Comment:
This feels not super in the spirit of the state machine. Thoughts, @jonahkagan?

Reply:
I think ideally `doScanWhenReady` would be replaced with a guarded transition ie. on L860

```
        ready_to_scan: {
          id: 'ready_to_scan',
          entry: [clearError, clearLastScan],
          invoke: pollPaperStatus(),
          on: {
            SCAN: 'scanning',
            SCANNER_NO_PAPER: 'no_paper',
            SCANNER_READY_TO_SCAN: [
              {
                target: 'scanning',
                // Only transition to 'scanning' if the guard (cond) evaluates to true
                cond: async (context, event) => await getCurrentAppFlowState({ ... }) === 'ballot:waiting_to_scan',
              },
              // do nothing if guarded transition fails
              { target: doNothing },
            ]
          },
        },
```

But it looks like async guards are not supported because there's not a good way to handle errors. There are various flavors of "make a new state instead" described [here](https://github.com/statelyai/xstate/issues/533#issuecomment-630627546). So something like 

```
          on: {
            ...
            SCANNER_READY_TO_SCAN: 'evaluate_should_start_scan',
        },
```

then `evaluate_should_start_scan` calls `await getCurrentAppFlowState` in its `entry` or `invoke` and transitions based on the result; basically just making the transition explicit instead of using an intermediate `SCAN` event. I _think_ that might result in ping ponging between `ready_to_scan` and `evaluate_should_start_scan`, so debatable if it's cleaner than the current implementation of `doScanWhenReady`.

Reply:
Yeah I think @kshen0's second suggestion about creating an intermediate "checking" state seems like the recommended solution based on [this comment](https://github.com/statelyai/xstate/issues/533#issuecomment-510459432) from the xstate library author. Ping-ponging would be the desired behavior while waiting - that seems right to me.

Reply:
Thanks for pushing on this and the suggestions for improvements. I am pretty happy with the end result: https://github.com/votingworks/vxsuite/pull/4403/commits/36035fee58d1ded90388257e84f0ed568156b893#diff-39166310c5456c6b6d7ed5a61720aeb48b89c2be78b81a4eeebdef3d981f3945R837-R851